### ISSUE TITLE
Probe file not linking with spike sorters

### ISSUE THREAD

**Issue by Erik2020-Neuronz** [SpikeInterface/spikeinterface#314]
Hello,

I am using Spike Interface to run neural data through multiple sorters simultaneously. These include Kilosort2, Spyking-Circus, and Klusta

While I can read my probe file into spike interface (m15_190315_145422.prb -> given at the bottom of this message) using the `load_probe_file` function, the output from kilosort does not order the channels as specified by the .prb file. Also, channel 0 was intentionally excluded from the probe file geometry but it still appeared in the phy output. 

In the phy output image (image 1_PhyDisplayKs2Output), we can see selected channels for a unit template. A similar spike waveforms appear in channel 12 and 14, both of which are spatially adjacent in the probe map (see image 2_ChannelGeometry). However, channel 13 does not contain the waveform, and is also not adjacent to 12 and 14 (see image 2). Despite this, the Phy output makes me think that it using 13 still being used for this template since it's displayed in the waveform view

my workflow is like so:

after specifying the data path in recording_folder, I then extract the data into `recording`

`recording = se.BinDatRecordingExtractor(recording_folder + '/m15_190315_145422.dat',30000,33,'int16')`

next I read in the probe file
`recording.load_probe_file('m15_190315_145422.prb')`
which gives the output
`<spikeextractors.subrecordingextractor.SubRecordingExtractor at 0x7f02848bcb70>`

I then read in my kilosort path
`ss.Kilosort2Sorter.set_kilosort2_path('path')`

and finally I run kilosort2 with the following:

`sorting_Ks2 = ss.run_kilosort2(recording, 
                              output_folder='secondaryTest_Ks2_5422', 
                              grouping_property='group', parallel=True, verbose=True)
print(f'Kilosot2 found {len(sorting_Ks2.get_unit_ids())} units')`

Is this the proper way to specify the probe file or do I need to take additional steps? Thank you 

I am running this through JupyterNotebook and am running an through an environment with python version 3.6.13 and running on Ubuntu version 20.04.

![PhyDisplayKs2Output](https://user-images.githubusercontent.com/79172753/137553765-824e335a-22c6-4676-88d8-8727399d6f9e.png)

![ChannelGeometry](https://user-images.githubusercontent.com/79172753/137553718-71f03680-5292-4c24-90be-324127a3747c.png)

m15_190315_145422.prb:
-------------------------------------------------------

total_nb_channels = 33
radius            = 100

channel_groups = {
    
   1: {'channels':[17,18,19,21,22,23,32,31,30,28,27,26,25,29,24,20,13,9,4,8,7,6,5,3,2,1,10,11,12,14,15,16],
       'graph' : [],
       'geometry': {           
            17: [  -18.0 ,  12.65],
            18: [  -18.0 ,  37.65],
            19: [  -18.0 ,  62.65],
            21: [  -18.0 ,  87.65],
            22: [  -18.0 ,  112.65],
            23: [  -18.0 ,  137.65],
            32: [  -18.0 ,  162.65],
            31: [  -18.0 ,  187.65],
            30: [  -18.0 ,  212.65],
            28: [  -18.0 ,  237.65],
            27: [  0.0 ,  0],
            26: [  0.0 ,  50.00],
            25: [  0.0 ,  100.00],
            29: [  0.0 ,  150.00],
            24: [  0.0 ,  200.00],
            20: [  0.0 ,  250.00],
            13: [  0.0 ,  275.00],
            9:  [  0.0 ,  225.00],
            4:  [  0.0 ,  175.00],
            8:  [  0.0 ,  125.00],
            7:  [  0.0 ,  75.00],
            6:  [  0.0 ,  25.00],
            5:  [  18.0 ,  237.65],
            3:  [  18.0 ,  212.65],
            2:  [  18.0 ,  187.65],
            1:  [  18.0 ,  162.65],
            10: [  18.0 ,  137.65],
            11: [  18.0 ,  112.65],
            12: [  18.0 ,  87.65],
            14: [  18.0 ,  62.65],
            15: [  18.0 ,  37.65],
            16: [  18.0 ,  12.65]}},
   
    }

**Comment by alejoe91** [comment]
Hi @Erik2020-Neuronz 

The `load_probe_file` function returns a NEW recording object. So you should do:

`recording = recording.load_probe_file('m15_190315_145422.prb')`

**Comment by alejoe91** [comment]
The 'problem' is that Phy requires a linear channel map, so we set a dummy one just for phy visualization. When you load the data back to SI you should have your original channel map back (we save it in a separate file)

**Comment by alejoe91** [comment]
@Erik2020-Neuronz can we close this?

**Comment by argunsah** [comment]
Hello,
I have a similar issue with the new API. When I create a probe as follows, the export to phy creates a 63 channel linear probe as exported object:

# generate a 8 shanks linear
probe0 = generate_linear_probe(num_elec=8, ypitch=100,
                               contact_shapes='square',
                               contact_shape_params={'width': 13.3})
probe1 = probe0.copy()
probe2 = probe0.copy()
probe3 = probe0.copy()
probe4 = generate_linear_probe(num_elec=7, ypitch=100,
                               contact_shapes='square',
                               contact_shape_params={'width': 13.3})
probe5 = probe0.copy()
probe6 = probe0.copy()
probe7 = probe0.copy()

probe1.move([200, 0])
probe2.move([400, 0])
probe3.move([600, 0])
probe4.move([800, 100])
probe5.move([1000, 0])
probe6.move([1200, 0])
probe7.move([1400, 0])

probe = combine_probes([probe0, probe1, probe2, probe3, probe4, probe5, probe6, probe7])

#channel_indices = np.arange(num_channels)
channel_indices = np.array([38,40,32,42,37,44,39,46,41,48,43,2,45,4,47,6,1,8,3,10,5,14,7,16,9,11,15,13,12,35,34,33,63,62,21,20,18,22,24,17,26,19,28,23,30,25,32,27,50,29,52,31,54,49,56,51,58,53,60,55,61,57,59])
channel_indices = channel_indices - 1 # zero python index system
probe.set_device_channel_indices(channel_indices)

however, if I create as follows, it creates a 8x8 probe:

# generate a 8 shanks linear
probe0 = generate_linear_probe(num_elec=8, ypitch=100,
                               contact_shapes='square',
                               contact_shape_params={'width': 13.3})
probe1 = probe0.copy()
probe2 = probe0.copy()
probe3 = probe0.copy()
probe4 = probe0.copy()
probe5 = probe0.copy()
probe6 = probe0.copy()
probe7 = probe0.copy()

probe1.move([200, 0])
probe2.move([400, 0])
probe3.move([600, 0])
probe4.move([800, 0])
probe5.move([1000, 0])
probe6.move([1200, 0])
probe7.move([1400, 0])

probe = combine_probes([probe0, probe1, probe2, probe3, probe4, probe5, probe6, probe7])

#channel_indices = np.arange(num_channels)
channel_indices = np.array([38,40,32,42,37,44,39,46,41,48,43,2,45,4,47,6,1,8,3,10,5,14,7,16,9,11,15,13,12,35,34,33,63,62,21,20,18,22,24,17,26,19,28,23,30,25,32,27,50,29,52,31,54,49,56,51,58,53,60,55,61,57,59,63])
channel_indices = channel_indices - 1 # zero python index system
probe.set_device_channel_indices(channel_indices)

Do you think there is a way of fixing this issue? (I use a 8x8 probe (neuronexus 8x8 edge), but we do not use channel 64, which in our case is the tip of shank 5. I couldn't figure out a different way of creating a 63 channel probe with a missing contact, hence the unusual probe shape). 

P.S. Probe shape is fine in plot and post-sorting cell locations visualized nicely. 

Best,

Ali

**Comment by samuelgarcia** [comment]
Hi,
You can do
```python
print(probe64) # should be 64 contacts
selection = ... # a vector with 63 indices
probe63 = probe64.get_slice(selection)
print(probe63) # should be 63 contacts
```

an other way to remove a channel is to do it at recording level
```python
recording = readXXXX()
recording64 = recordingset_probe(probe64)
print(recording64)
print(recording64.channel_ids)
seom_channel_ids = ... # a vector with channel ids you want to keep
recording63 = recording64.channel_slice(some_channel_ids)
# here the probe is internally sliced
print(recording63)

```

**Comment by argunsah** [comment]
Hello,

I tried your suggestion. I get an AssertionError.

I used the following code: 

# generate a 8 shanks linear
probe0 = generate_linear_probe(num_elec=8, ypitch=100,
                               contact_shapes='square',
                               contact_shape_params={'width': 13.3})
probe1 = probe0.copy()
probe2 = probe0.copy()
probe3 = probe0.copy()
probe4 = probe0.copy()
probe5 = probe0.copy()
probe6 = probe0.copy()
probe7 = probe0.copy()

probe1.move([200, 0])
probe2.move([400, 0])
probe3.move([600, 0])
probe4.move([800, 0])
probe5.move([1000, 0])
probe6.move([1200, 0])
probe7.move([1400, 0])

probe = combine_probes([probe0, probe1, probe2, probe3, probe4, probe5, probe6, probe7])
#channel_indices = np.arange(num_channels)
channel_indices = np.array([38,40,32,42,37,44,39,46,41,48,43,2,45,4,47,6,1,8,3,10,5,14,7,
16,9,11,15,13,12,35,34,33,64,63,62,21,20,18,22,24,17,26,19,28,23,30,25,32,27,50,29,52,31,54,
49,56,51,58,53,60,55,61,57,59])
channel_indices = channel_indices - 1 # zero python index system
probe.set_device_channel_indices(channel_indices)

print(probe)

selection = np.array([38,40,32,42,37,44,39,46,41,48,43,2,45,4,47,6,1,8,3,10,5,14,7,
16,9,11,15,13,12,35,34,33,63,62,21,20,18,22,24,17,26,19,28,23,30,25,32,27,50,29,52,31,54,
49,56,51,58,53,60,55,61,57,59])
selection = selection - 1
#selection = np.ones((64, 1), dtype=bool)
#selection[33,0]=0
print(selection)
probe = probe.get_slice(selection)

**Comment by samuelgarcia** [comment]
Hi,
`selection` is not the `channel_indices` at all.
In my example `selection` refer to a list or a vector of the contact you want to keep.

For instance
```python
selection = list(range(0, 12)) + list(range(13, 64))
```
if you want to remove the contact 12 (0 based)

**Comment by argunsah** [comment]
Hello,
This suggestion worked perfectly, thank you!

However, my main issue persists. This way of creating the probe did not solve the probeview issue in phy issue of creating a linear 64 channel probe instead of the one I created. 

(I get this error: 10:43:21.892 [E] model:389            Some channels are on the same position, please check the channel positions file.)

Probe Looks nice:
![spikeLocations](https://user-images.githubusercontent.com/10572737/156539326-7f5b6f86-44a2-4c71-ba32-56e4202888c1.png)

But error happens and this is how it looks in phy.
![Screenshot from 2022-03-03 10-43-58](https://user-images.githubusercontent.com/10572737/156539394-cefc2840-4dbe-48b5-ab72-769dad1edf73.png)

Thank you for the help!

**Comment by samuelgarcia** [comment]
I don't know phy very well.
@alejoe91 : could you help about this lenar stuff in phy.


@argunsah : you should try [spikeinterface-gui](https://github.com/SpikeInterface/spikeinterface-gui) !!
No need to export to phy and quite good viewer also but no manual curation.

**Comment by alejoe91** [comment]
@argunsah can you paste here the output of `print(recording.get_channel_locations()`? Let's make sure that all locations are unique!

**Comment by alejoe91** [comment]
Similarly, can you try to load and print the `channel_positions.npy` in the phy folder? They should correspond to the ones printed form spikeinterface

**Comment by argunsah** [comment]
Hello,
Channel position [1000,700] is repeating twice. I am deleting electrode 33 and the repeated ones have indexes 31 and 32, it should have something to the with that I am guessing.

        x      y contact_shapes  width shank_ids contact_ids
0      0.0    0.0         square   13.3         0            
1      0.0  100.0         square   13.3         0            
2      0.0  200.0         square   13.3         0            
3      0.0  300.0         square   13.3         0            
4      0.0  400.0         square   13.3         0            
..     ...    ...            ...    ...       ...         ...
58  1400.0  300.0         square   13.3         7            
59  1400.0  400.0         square   13.3         7            
60  1400.0  500.0         square   13.3         7            
61  1400.0  600.0         square   13.3         7            
62  1400.0  700.0         square   13.3         7            

[63 rows x 6 columns]
['0' '0' '0' '0' '0' '0' '0' '0' '1' '1' '1' '1' '1' '1' '1' '1' '2' '2'
 '2' '2' '2' '2' '2' '2' '3' '3' '3' '3' '3' '3' '3' '3' '4' '4' '4' '4'
 '4' '4' '4' '5' '5' '5' '5' '5' '5' '5' '5' '6' '6' '6' '6' '6' '6' '6'
 '6' '7' '7' '7' '7' '7' '7' '7' '7']
[[ 400.    0.]
 [ 200.  300.]
 [ 400.  200.]
 [ 200.  500.]
 [ 400.  400.]
 [ 200.  700.]
 [ 400.  600.]
 [ 400.  100.]
 [ 600.    0.]
 [ 400.  300.]
 [ 600.  100.]
 [ 600.  400.]
 [ 600.  300.]
 [ 400.  500.]
 [ 600.  200.]
 [ 400.  700.]
 [1000.    0.]
 [ 800.  500.]
 [1000.  200.]
 [ 800.  400.]
 [ 800.  300.]
 [ 800.  600.]
 [1000.  400.]
 [ 800.  700.]
 [1000.  600.]
 [1000.  100.]
 [1200.    0.]
 [1000.  300.]
 [1200.  200.]
 [1000.  500.]
 [1200.  400.]
 [1000.  700.]
 [1000.  700.]
 [ 600.  700.]
 [ 600.  600.]
 [ 600.  500.]
 [   0.  400.]
 [   0.    0.]
 [   0.  600.]
 [   0.  100.]
 [ 200.    0.]
 [   0.  300.]
 [ 200.  200.]
 [   0.  500.]
 [ 200.  400.]
 [   0.  700.]
 [ 200.  600.]
 [ 200.  100.]
 [1200.  600.]
 [1200.  100.]
 [1400.    0.]
 [1200.  300.]
 [1400.  200.]
 [1200.  500.]
 [1400.  400.]
 [1200.  700.]
 [1400.  600.]
 [1400.  100.]
 [1400.  700.]
 [1400.  300.]
 [1400.  500.]
 [ 800.  200.]
 [ 800.  100.]]

**Comment by alejoe91** [comment]
Yes definitely that is the problem! Can you try to investigate when this duplication is happening?

**Comment by argunsah** [comment]
This is what I am doing exactly:

# generate a 8 shanks linear
probe0 = generate_linear_probe(num_elec=8, ypitch=100,
                               contact_shapes='square',
                               contact_shape_params={'width': 13.3})
probe1 = probe0.copy()
probe2 = probe0.copy()
probe3 = probe0.copy()
probe4 = probe0.copy()
probe5 = probe0.copy()
probe6 = probe0.copy()
probe7 = probe0.copy()

probe1.move([200, 0])
probe2.move([400, 0])
probe3.move([600, 0])
probe4.move([800, 0])
probe5.move([1000, 0])
probe6.move([1200, 0])
probe7.move([1400, 0])

probe = combine_probes([probe0, probe1, probe2, probe3, probe4, probe5, probe6, probe7])

channel_indices = np.array([38,40,32,42,37,44,39,46,41,48,43,2,45,4,47,6,1,8,3,10,5,14,7,
16,9,11,15,13,12,35,34,33,64,63,62,21,20,18,22,24,17,26,19,28,23,30,25,32,27,50,29,52,31,54,
49,56,51,58,53,60,55,61,57,59])
channel_indices = channel_indices - 1 # zero python index system
probe.set_device_channel_indices(channel_indices)

print(probe)

selection = list(range(0, 32)) + list(range(33, 64))
print(selection)
probe = probe.get_slice(selection)
print(probe.device_channel_indices)

**Comment by alejoe91** [comment]
Can you do `len(np.unique(channel_indices))`? @samuelgarcia can you try to reproduce the issue?

**Comment by argunsah** [comment]
print(len(np.unique(channel_indices)))
gives
63

which is correct number of channels

I will try to fix channel locations manually to see if it will solve the issue

**Comment by argunsah** [comment]
Hello,
I solved the issue. It was my mistake. I repeated channel 32 twice but then I deleted channel 32 with get_slice (repeated channel and deleted channel having same index in coincidence) which somehow created the error while keeping total number of electrodes 63. Now in the fixed version total print(len(np.unique(channel_indices))) command gives 64 (so it actually gives the total number of channels before the removal of one channel I removed with get_slice).

So, it was my mistake but it was a funky mistake to debug due to the coincidences.

Thank you for all the help. All works in phy and spikeinterface_gui as well!
